lib/lib-dynamodb/src/DynamoDBDocumentClient.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  DynamoDBClient,
+  DynamoDBClientResolvedConfig,
+  ServiceInputTypes,
+  ServiceOutputTypes,
+} from "@aws-sdk/client-dynamodb";
+import { Client as __Client } from "@aws-sdk/smithy-client";
+import { HttpHandlerOptions as __HttpHandlerOptions } from "@aws-sdk/types";
+import { marshallOptions, unmarshallOptions } from "@aws-sdk/util-dynamodb";
+
+export type TranslateConfig = {
+  marshallOptions?: marshallOptions;
+  unmarshallOptions?: unmarshallOptions;
+};
+
+export type DynamoDBDocumentClientResolvedConfig = DynamoDBClientResolvedConfig & {
+  translateConfig?: TranslateConfig;
+};
+
+/**
+ * The document client simplifies working with items in Amazon DynamoDB by
+ * abstracting away the notion of attribute values. This abstraction annotates native
+ * JavaScript types supplied as input parameters, as well as converts annotated
+ * response data to native JavaScript types.
+ *
+ * ## Marshalling Input and Unmarshalling Response Data
+ *
+ * The document client affords developers the use of native JavaScript types
+ * instead of `AttributeValue`s to simplify the JavaScript development
+ * experience with Amazon DynamoDB. JavaScript objects passed in as parameters
+ * are marshalled into `AttributeValue` shapes required by Amazon DynamoDB.
+ * Responses from DynamoDB are unmarshalled into plain JavaScript objects
+ * by the `DocumentClient`. The `DocumentClient` does not accept
+ * `AttributeValue`s in favor of native JavaScript types.
+ *
+ * |          JavaScript Type          | DynamoDB AttributeValue |
+ * | :-------------------------------: | ----------------------- |
+ * |              String               | S                       |
+ * |          Number / BigInt          | N                       |
+ * |              Boolean              | BOOL                    |
+ * |               null                | NULL                    |
+ * |               Array               | L                       |
+ * |              Object               | M                       |
+ * |                Set                | BS / NS / SS            |
+ * | Uint8Array, Buffer, File, Blob... | B                       |
+ *
+ * ### Example
+ *
+ * Here is an example list which is sent to DynamoDB client in an operation:
+ *
+ * ```json
+ * { "L": [{ "NULL": true }, { "BOOL": false }, { "N": 1 }, { "S": "two" }] }
+ * ```
+ *
+ * The DynamoDB document client abstracts the attribute values as follows in
+ * both input and output:
+ *
+ * ```json
+ * [null, false, 1, "two"]
+ * ```
+ *
+ * @see {@link https://www.npmjs.com/package/@aws-sdk/client-dynamodb @aws-sdk/client-dynamodb}
+ */
+export class DynamoDBDocumentClient extends __Client<
+  __HttpHandlerOptions,
+  ServiceInputTypes,
+  ServiceOutputTypes,
```

Comment:
This service input and output types should be the doc client's input and output types that already overrides the basic client's type, *not* the basic client's types. Now it seems to work even if you supply the doc client's commands. It's a little concerning that doc client also can send basic ddb client's commands without type safety.

Reply:
I guess it may have something to do with the fact that input types are `type` not `interface`. TypeScript has issue with less type safety for `type`. If that's the case we need to follow up by updating the input and output types to interface.
https://github.com/aws/aws-sdk-js-v3/blob/9c21f14412f2b1f591422f3c67dedbe886db723b/clients/client-dynamodb/commands/BatchExecuteStatementCommand.ts#L20

Reply:
Fixed in [`4b26208` (#2097)](https://github.com/aws/aws-sdk-js-v3/pull/2097/commits/4b2620834c3147f6520a89fb9f54954874193875)